File: app/src/main/java/soup/movie/settings/TheaterSetting.java
package soup.movie.settings;

import android.content.SharedPreferences;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.apache.commons.collections4.ListUtils;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import soup.movie.data.model.Theater;

public class TheaterSetting {

    private static final String KEY = "favorite_theaters";
    private static final String DEFAULT_VALUE = "";

    private final SharedPreferences preferences;

    public TheaterSetting(SharedPreferences preferences) {
        this.preferences = preferences;
    }

    public List<Theater> getFavoriteTheaters() {
        return ListUtils.emptyIfNull(fromJson(preferences.getString(KEY, DEFAULT_VALUE)));
    }

    public void setFavoriteTheaters(List<Theater> theaters) {
        preferences.edit().putString(KEY, toJson(theaters)).apply();
    }

    private static String toJson(List<Theater> theaterCodeList) {
        return new Gson().toJson(theaterCodeList);
    }

    private static List<Theater> fromJson(String jsonStr) {
        Type theaterListType = new TypeToken<ArrayList<Theater>>(){}.getType();
        return new Gson().fromJson(jsonStr, theaterListType);
    }
}

File: app/src/main/java/soup/movie/settings/TheaterSetting.kt
package soup.movie.settings

import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import soup.movie.data.model.Theater

class TheaterSetting(preferences: SharedPreferences) :
        PrefSetting<List<Theater>>(preferences) {

    override fun getDefaultValue(preferences: SharedPreferences): List<Theater> {
        return fromJson(preferences.getString(KEY, DEFAULT_VALUE))
    }

    override fun saveValue(preferences: SharedPreferences, value: List<Theater>) {
        return preferences.edit().putString(KEY, toJson(value)).apply()
    }

    companion object {

        private const val KEY = "favorite_theaters"
        private const val DEFAULT_VALUE = ""

        private val type = object : TypeToken<ArrayList<Theater>>() {}.type

        private fun toJson(theaterCodeList: List<Theater>): String {
            return Gson().toJson(theaterCodeList)
        }

        private fun fromJson(jsonStr: String): List<Theater> {
            return Gson().fromJson<List<Theater>>(jsonStr, type).orEmpty()
        }
    }
}
